crates/allocator/src/bump.rs

```diff
@@ -0,0 +1,243 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple bump allocator.
+//!
+//! Its goal to have a much smaller footprint than the admittedly more full-featured `wee_alloc`
+//! allocator which is currently being used by ink! smart contracts.
+//!
+//! The heap which is used by this allocator is built from pages of Wasm memory (each page is `64KiB`).
+//! We will request new pages of memory as needed until we run out of memory, at which point we
+//! will crash with an `OOM` error instead of freeing any memory.
+
+use core::alloc::{
+    GlobalAlloc,
+    Layout,
+};
+
+/// A page in Wasm is `64KiB`
+const PAGE_SIZE: usize = 64 * 1024;
+
+static mut INNER: InnerAlloc = InnerAlloc::new();
+
+/// A bump allocator suitable for use in a Wasm environment.
+pub struct BumpAllocator;
+
+unsafe impl GlobalAlloc for BumpAllocator {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        match INNER.alloc(layout) {
+            Some(start) => start as *mut u8,
+            None => core::ptr::null_mut(),
+        }
+    }
+
+    #[inline]
+    unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+        // A new page in Wasm is guaranteed to already be zero initialized, so we can just use our
+        // regular `alloc` call here and save a bit of work.
+        self.alloc(layout)
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+}
+
+#[cfg_attr(feature = "std", derive(Debug, Copy, Clone))]
+struct InnerAlloc {
+    /// Points to the start of the next available allocation.
+    next: usize,
+
+    /// The address of the upper limit of our heap.
+    upper_limit: usize,
+
+    /// The number of page requests made so far.
+    ///
+    /// This is meant to mimic the behavior exhibited by `core::arch::wasm32::memory_grow`
+    #[cfg(feature = "std")]
+    page_requests: usize,
+}
+
+impl InnerAlloc {
+    const fn new() -> Self {
+        Self {
+            next: 0,
+            upper_limit: 0,
+            #[cfg(feature = "std")]
+            page_requests: 0,
+        }
+    }
+
+    cfg_if::cfg_if! {
+        if #[cfg(all(not(feature = "std"), target_arch = "wasm32"))] {
+            /// Request a `pages` number of pages of Wasm memory. Each page is `64KiB` in size.
+            ///
+            /// Returns `None` if a page is not available.
+            fn request_pages(&mut self, pages: usize) -> Option<usize> {
+                let prev_page = core::arch::wasm32::memory_grow(0, pages);
+                if prev_page == usize::MAX {
+                    return None;
+                }
+
+                prev_page.checked_mul(PAGE_SIZE)
+            }
+
+        } else if #[cfg(feature = "std")] {
```

Comment:
I would rather see `[cfg(test)]` being used here. It is simpler and expressed the intent of the code. Also because the implementation is unsound for everything but inspecting the internal state via the unit tests.

Reply:
So we can't _just_ use `cfg[test]` here, otherwise it won't compile. I've changed it to be `#[cfg(any(test, feature = "std"))]` so that at least there's an in-code reference that this is meant to be used for testing

Reply:
> So we can't just use cfg[test] here, otherwise it won't compile.

Any idea why that is? I only see this used in tests outside of `no_std`.

Reply:
When the `cargo test` runs it does two passes. One with `--features default std`, and another with `<previous flags> --test` (you can verify this by running with `--verbose`).

This means that on the one pass there's no function called `request_pages` unless the `std` feature is enabled, which leads to compilation failures.

Reply:
But that compilation pass with `std` enabled and `test` disabled should never be executed. So you could add a branch for that where you just put `unreachable!("This code should only be ever called by tests")`.